com.hazelcast.client.internal.metrics.ClientJmxMetricsTest.testNoMBeanLeak

_4.2.z_ (commit e22aab3cffbb36863a91276dd2506b4fcf4a7f6c)

Failed on Windows with Oracle JDK 8: http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-4.maintenance-Windows-OracleJDK8/301/testReport/com.hazelcast.client.internal.metrics/ClientJmxMetricsTest/testNoMBeanLeak/

Stacktrace:
```
java.lang.AssertionError: expected:<0> but was:<45>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at com.hazelcast.internal.metrics.jmx.JmxPublisherTestHelper.assertNoMBeans(JmxPublisherTestHelper.java:60)
	at com.hazelcast.client.internal.metrics.ClientJmxMetricsTest.testNoMBeanLeak(ClientJmxMetricsTest.java:52)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

Standard output:
```
19:15:34,406  INFO |testNoMBeanLeak| - [JmxPublisherTestHelper] testNoMBeanLeak - Dangling metrics MBeans created by 22332@EC2AMAZ-IK66GLI: [Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=tcp.outputThread,tag0="thread=hz.hz.client_1753.IO.thread-out-0"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=tcp.connection.in,tag0="pipelineId=/0.0.0.0:57074->null"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=gc], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=tcp.inputThread,tag0="thread=hz.hz.client_1754.IO.thread-in-0"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=invocations], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=tcp.balancer], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=runtime], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=tcp.balancer], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=tcp.balancer], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=runtime], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=listeners], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=file.partition,tag0="dir=user.home"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=os], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=tcp], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=tcp.outputThread,tag0="thread=hz.hz.client_1754.IO.thread-out-0"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=listeners], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=tcp.connection.out,tag0="pipelineId=/0.0.0.0:57074->null"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=classloading], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=memory], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=classloading], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=invocations], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=classloading], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=file.partition,tag0="dir=user.home"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=file.partition,tag0="dir=user.home"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=gc], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=tcp], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=gc], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=invocations], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=tcp.inputThread,tag0="thread=hz.hz.client_1753.IO.thread-in-0"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=tcp.connection.in,tag0="pipelineId=/0.0.0.0:57071->null"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=memory], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=thread], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=tcp.connection.in,tag0="pipelineId=/0.0.0.0:57073->null"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=thread], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=memory], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=os], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=tcp], Metric[com.hazelcast:type=Metrics,instance=hz.client_1754,prefix=os], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=tcp.connection.out,tag0="pipelineId=/0.0.0.0:57071->null"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=tcp.outputThread,tag0="thread=hz.hz.client_1752.IO.thread-out-0"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=tcp.inputThread,tag0="thread=hz.hz.client_1752.IO.thread-in-0"], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=thread], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=runtime], Metric[com.hazelcast:type=Metrics,instance=hz.client_1753,prefix=listeners], Metric[com.hazelcast:type=Metrics,instance=hz.client_1752,prefix=tcp.connection.out,tag0="pipelineId=/0.0.0.0:57073->null"]]
19:15:34,644  INFO |autoDetectionNotUsedWhenOtherDiscoveryEnabled| - [ClientConnectionManager] hz.client_1753.internal-4 - hz.client_1753 [dev] [4.2.1-SNAPSHOT] Trying to connect to Member [127.0.0.1]:5710 - 8ee4f5a7-5561-4005-8592-bc8eaae98784
19:15:34,645  WARN |defaultDiscovery| - [ClientConnectionManager] hz.client_1754.internal-9 - hz.client_1754 [dev] [4.2.1-SNAPSHOT] Exception during initial connection to Member [127.0.0.1]:5701 - 4ecef362-b7da-4425-a276-c1d8c90a3646: com.hazelcast.core.HazelcastException: java.io.IOException: Connection refused: no further information to address /127.0.0.1:5701
19:15:34,645  WARN |autoDetectionDisabled| - [ClientConnectionManager] hz.client_1752.internal-4 - hz.client_1752 [dev] [4.2.1-SNAPSHOT] Exception during initial connection to Member [127.0.0.1]:5701 - 4ecef362-b7da-4425-a276-c1d8c90a3646: com.hazelcast.core.HazelcastException: java.io.IOException: Connection refused: no further information to address /127.0.0.1:5701
19:15:34,645  INFO |defaultDiscovery| - [ClientConnectionManager] hz.client_1754.internal-9 - hz.client_1754 [dev] [4.2.1-SNAPSHOT] Trying to connect to [127.0.0.1]:5703
19:15:34,645  INFO |autoDetectionDisabled| - [ClientConnectionManager] hz.client_1752.internal-4 - hz.client_1752 [dev] [4.2.1-SNAPSHOT] Trying to connect to [127.0.0.1]:5702
BuildInfo right after testNoMBeanLeak(com.hazelcast.client.internal.metrics.ClientJmxMetricsTest): BuildInfo{version='4.2.1-SNAPSHOT', build='20210519', buildNumber=20210519, revision=e22aab3, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'testNoMBeanLeak(com.hazelcast.client.internal.metrics.ClientJmxMetricsTest):'
19:15:30, accumulated pauses: 863 ms, max pause: 237 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/test/java/com/hazelcast/client/internal/metrics/ClientJmxMetricsTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/client/internal/metrics/ClientJmxMetricsTest.java
@@ -49,7 +49,7 @@ public class ClientJmxMetricsTest extends HazelcastTestSupport {
 
     @Test
     public void testNoMBeanLeak() {
-        helper.assertNoMBeans();
+        helper.clearMBeans();
 
         ClientConfig clientConfig = new ClientConfig();
         clientConfig.getMetricsConfig()
@@ -62,11 +62,6 @@ public class ClientJmxMetricsTest extends HazelcastTestSupport {
         assertTrueEventually(() -> helper.assertMBeanContains(expectedObjectName));
 
         client.shutdown();
-        // if failing, grep for "test-jvm"
-        // there should be a Metric[com.hazelcast:type=Metrics,instance=$CLIENT_NAME,prefix=test-jvm-$JVM_NAME] metric if the
-        // client was created with TestHazelcastFactory
-        // $JVM_NAME contains the PID, so $CLIENT_NAME + $JVM_NAME together helps identifying the test left the given metric
-        // around
         helper.assertNoMBeans();
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/internal/metrics/jmx/JmxPublisherTestHelper.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/metrics/jmx/JmxPublisherTestHelper.java
@@ -18,6 +18,7 @@ package com.hazelcast.internal.metrics.jmx;
 
 import com.hazelcast.internal.metrics.jmx.MetricsMBean.Type;
 import com.hazelcast.internal.util.BiTuple;
+import com.hazelcast.internal.util.ExceptionUtil;
 import com.hazelcast.internal.util.TriTuple;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.logging.Logger;
@@ -49,6 +50,20 @@ public class JmxPublisherTestHelper {
         objectNameNoModule = new ObjectName(domainPrefix + ":*");
         objectNameWithModule = new ObjectName(domainPrefix + "." + MODULE_NAME + ":*");
         platformMBeanServer = ManagementFactory.getPlatformMBeanServer();
+    }
+
+    public void clearMBeans() {
+        queryOurInstances().forEach(this::clearMBean);
+        assertNoMBeans();
+    }
+
+    private void clearMBean(ObjectInstance instance) {
+        try {
+            ObjectName name = instance.getObjectName();
+            platformMBeanServer.unregisterMBean(name);
+        } catch (Exception e) {
+            throw ExceptionUtil.rethrow(e);
+        }
     }
 
     public void assertNoMBeans() {

--- a/hazelcast/src/test/java/com/hazelcast/internal/metrics/jmx/MemberJmxMetricsTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/metrics/jmx/MemberJmxMetricsTest.java
@@ -50,7 +50,7 @@ public class MemberJmxMetricsTest {
 
     @Test
     public void testNoMBeanLeak() {
-        helper.assertNoMBeans();
+        helper.clearMBeans();
 
         Config config = smallInstanceConfig();
         config.getMetricsConfig()

